c10/core/impl/cow/state.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <c10/core/impl/cow/simulator.h>
+#include <c10/macros/Macros.h>
+#include <c10/util/Optional.h>
+#include <c10/util/intrusive_ptr.h>
+
+#include <cstdint>
+#include <mutex>
+
+namespace c10::impl::cow {
+
+// Responsible for managing the copy-on-write simulation state
+// machine.
```

Comment:
An abstract description of the state machine here would be helpful

Reply:
apparently the state machine where you stop tracking after a warning has been raised. But not clear if this is worth implementing?

Reply:
Agreed. That will come in the next iteration. I'm going to formally define an enum with the states and transitions and document each individually.

The null state is important. I want to drop the cost of the null state to an atomic read. In the common case, we can make it an uncontested read that will indicate that there's no lazy copies to be concerned about. This will be much better than the write to acquire a mutex.

Reply:
OK, the state machine has manifested itself.